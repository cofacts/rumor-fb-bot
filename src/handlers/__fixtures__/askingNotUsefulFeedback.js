export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
  },
};
