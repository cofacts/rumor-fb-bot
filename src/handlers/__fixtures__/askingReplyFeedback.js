export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};

export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const articleData = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};
