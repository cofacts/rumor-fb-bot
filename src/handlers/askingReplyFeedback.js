import { t, ngettext, msgid } from 'ttag';
import gql from '../gql';
import ga from '../ga';
import {
  getArticleURL,
  createPostbackAction,
  createTypeWords,
  ellipsis,
} from './utils';

/**
 * The state that user has given feedback about the article and replies we displayed.
 * We ask them to send other people our replies if they're satisfied or tell us why
 * they think the replies are not useful.
 */
export default async function askingReplyFeedback(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;

  if (!data.selectedReplyId) {
    throw new Error('selectedReply not set in data');
  }

  const visitor = ga('FB-' + userId, data.selectedArticleText);
  visitor.screenview({ screenName: state });

  // Track when user give feedback.
  visitor.event({
    ec: 'UserInput',
    ea: 'Feedback-Vote',
    el: `${data.selectedArticleId}/${data.selectedReplyId}`,
  });

  if (event.input === 'y') {
    const {
      data: {
        action: { feedbackCount },
      },
    } = await gql`
      mutation($vote: FeedbackVote!, $articleId: String!, $replyId: String!) {
        action: CreateOrUpdateArticleReplyFeedback(
          vote: $vote
          articleId: $articleId
          replyId: $replyId
        ) {
          feedbackCount
        }
      }
    `(
      {
        articleId: data.selectedArticleId,
        replyId: data.selectedReplyId,
        vote: 'UPVOTE',
      },
      { userId }
    );
    const {
      data: { GetReply },
    } = await gql`
      query($replyId: String!) {
        GetReply(id: $replyId) {
          type
          text
          reference
        }
      }
    `({
      replyId: data.selectedReplyId,
    });

    const articleUrl = getArticleURL(data.selectedArticleId);

    // FIXME: A string is used for length here because of the need of i18n.
    // The ideal length of ellipsis displayed may differ depending on the language.
    // For example, chinese characters are wider so we need a shorter length.
    const ellipsisLength = 30;
    const content = ellipsis(data.selectedArticleText, ellipsisLength);
    const contentType = createTypeWords(GetReply.type);
    const sharedContent = {
      title: t`Hey someone else says ‚Äú${content}‚Äù is ${contentType}!`,
      subtitle: t`Please refer to ${articleUrl} for other replies to this message and references!`,
      buttons: [
        {
          type: 'web_url',
          url: articleUrl,
          title: t`See other replies`,
        },
      ],
    };

    const otherFeedbackCount = feedbackCount - 1;
    replies = [
      {
        type: 'text',
        content: {
          text:
            otherFeedbackCount > 0
              ? ngettext(
                  msgid`We've received feedback from you and ${otherFeedbackCount} other user!`,
                  `We've received feedback from you and ${otherFeedbackCount} other users!`,
                  otherFeedbackCount
                )
              : t`Thanks. You're the first one who gave feedback on this reply!`,
        },
      },
      {
        type: 'generic',
        content: {
          attachment: {
            type: 'template',
            payload: {
              template_type: 'button',
              text:
                'üì≤ ' +
                t`Don't forget to forward the messages above to others and share with them!` +
                '\nüíÅ ' +
                t`And feel free to submit your own reply if you have anything to say about this!`,
              buttons: [
                {
                  type: 'element_share',
                  share_contents: {
                    attachment: {
                      type: 'template',
                      payload: {
                        template_type: 'generic',
                        elements: [sharedContent],
                      },
                    },
                  },
                },
                {
                  type: 'web_url',
                  url: articleUrl,
                  title: t`Submit a new reply`,
                },
              ],
            },
          },
        },
      },
    ];

    state = '__INIT__';
    visitor.send();
    return { data, state, event, issuedAt, userId, replies, isSkipUser };
  } else if (event.input === 'n') {
    replies = [
      {
        type: 'buttons',
        content: {
          attachment: {
            type: 'template',
            payload: {
              template_type: 'button',
              text: t`Why do you find this reply not helpful? Please tell us in messages. If you want to skip this, click the skip button below.`,
              buttons: [createPostbackAction(t`Skip`, 'n')],
            },
          },
        },
      },
    ];

    state = 'ASKING_NOT_USEFUL_FEEDBACK';
    visitor.send();
  } else {
    replies = [];
  }
  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
