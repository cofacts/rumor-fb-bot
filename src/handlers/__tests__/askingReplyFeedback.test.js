jest.mock('../../gql');

import askingReplyFeedback from '../askingReplyFeedback';
import * as apiResult from '../__fixtures__/askingReplyFeedback';
import gql from '../../gql';

const commonParamsYes = {
  data: {
    searchedText: '貼圖',
    foundArticleIds: [
      'AWDZYXxAyCdS-nWhumlz',
      '5483323992880-rumor',
      'AV-Urc0jyCdS-nWhuity',
      'AVsh8u7StKp96s659Dgq',
    ],
    selectedArticleId: 'AWDZYXxAyCdS-nWhumlz',
    selectedArticleText:
      '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    foundReplyIds: ['AWDZeeV0yCdS-nWhuml8'],
    selectedReplyId: 'AWDZeeV0yCdS-nWhuml8',
  },
  state: 'ASKING_REPLY_FEEDBACK',
  event: {
    type: 'postback',
    input: 'y',
    timestamp: 1519019734813,
    postback: { data: '{"input":"y","issuedAt":1519019701265}' },
  },
  issuedAt: 1519019735467,
  userId: 'Uaddc74df8a3a176b901d9d648b0fc4fe',
  replies: undefined,
  isSkipUser: false,
};

it('handles "yes" postback with no other existing feedbacks', async () => {
  gql.__push(apiResult.oneFeedback);
  gql.__push(apiResult.articleData);
  expect(await askingReplyFeedback(commonParamsYes)).toMatchSnapshot();
});

it('handles "yes" postback with other existing feedbacks', async () => {
  gql.__push(apiResult.twoFeedbacks);
  gql.__push(apiResult.articleData);
  expect(await askingReplyFeedback(commonParamsYes)).toMatchSnapshot();
});

const commonParamsNo = {
  data: {
    searchedText: '貼圖',
    foundArticleIds: [
      'AWDZYXxAyCdS-nWhumlz',
      '5483323992880-rumor',
      'AV-Urc0jyCdS-nWhuity',
      'AVsh8u7StKp96s659Dgq',
    ],
    selectedArticleId: 'AWDZYXxAyCdS-nWhumlz',
    foundReplyIds: ['AWDZeeV0yCdS-nWhuml8'],
    selectedReplyId: 'AWDZeeV0yCdS-nWhuml8',
  },
  state: 'ASKING_REPLY_FEEDBACK',
  event: {
    type: 'postback',
    input: 'n',
    timestamp: 1519019734813,
    postback: { data: '{"input":"n","issuedAt":1519019701265}' },
  },
  issuedAt: 1519019735467,
  userId: 'Uaddc74df8a3a176b901d9d648b0fc4fe',
  replies: undefined,
  isSkipUser: false,
};

it('handles "no" postback with other existing feedbacks', async () => {
  gql.__push(apiResult.oneFeedback);

  expect(await askingReplyFeedback(commonParamsNo)).toMatchSnapshot();
});

afterEach(() => {
  gql.__reset();
});
