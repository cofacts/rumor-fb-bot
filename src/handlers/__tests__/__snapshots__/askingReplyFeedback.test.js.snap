// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles "no" postback with other existing feedbacks 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AWDZYXxAyCdS-nWhumlz",
      "5483323992880-rumor",
      "AV-Urc0jyCdS-nWhuity",
      "AVsh8u7StKp96s659Dgq",
    ],
    "foundReplyIds": Array [
      "AWDZeeV0yCdS-nWhuml8",
    ],
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": "n",
    "postback": Object {
      "data": "{\\"input\\":\\"n\\",\\"issuedAt\\":1519019701265}",
    },
    "timestamp": 1519019734813,
    "type": "postback",
  },
  "isSkipUser": false,
  "issuedAt": 1519019735467,
  "replies": Array [
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "buttons": Array [
              Object {
                "payload": "n",
                "title": "Skip",
                "type": "postback",
              },
            ],
            "template_type": "button",
            "text": "Why do you find this reply not helpful? Please tell us in messages. If you want to skip this, click the skip button below.",
          },
          "type": "template",
        },
      },
      "type": "buttons",
    },
  ],
  "state": "ASKING_NOT_USEFUL_FEEDBACK",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;

exports[`handles "yes" postback with no other existing feedbacks 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AWDZYXxAyCdS-nWhumlz",
      "5483323992880-rumor",
      "AV-Urc0jyCdS-nWhuity",
      "AVsh8u7StKp96s659Dgq",
    ],
    "foundReplyIds": Array [
      "AWDZeeV0yCdS-nWhuml8",
    ],
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedArticleText": "(0)(1)(/)(0)(9)(line)免費貼圖

「[全螢幕貼圖]生活市集x生活小黑熊」
https://line.me/S/sticker/10299

「松果購物 x 狗與鹿(動態貼圖) 」
https://line.me/S/sticker/10300

「LINE TV X 我的男孩限免貼圖」
https://line.me/S/sticker/10207",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": "y",
    "postback": Object {
      "data": "{\\"input\\":\\"y\\",\\"issuedAt\\":1519019701265}",
    },
    "timestamp": 1519019734813,
    "type": "postback",
  },
  "isSkipUser": false,
  "issuedAt": 1519019735467,
  "replies": Array [
    Object {
      "content": Object {
        "text": "Thanks. You are the first one who gave feedback on this reply :)",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "buttons": Array [
              Object {
                "share_contents": Object {
                  "attachment": Object {
                    "payload": Object {
                      "elements": Array [
                        Object {
                          "buttons": Array [
                            Object {
                              "title": "See other replies",
                              "type": "web_url",
                              "url": "https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz",
                            },
                          ],
                          "subtitle": "Please refer to https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz for other replies to this message and references!",
                          "title": "Hey someone else says \\"(0)(1)(/)(0)(9)(line)免費貼圖

「⋯⋯ is ⚠️️ OFF TOPIC!",
                        },
                      ],
                      "template_type": "generic",
                    },
                    "type": "template",
                  },
                },
                "type": "element_share",
              },
              Object {
                "title": "Submit a new reply",
                "type": "web_url",
                "url": "https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz",
              },
            ],
            "template_type": "button",
            "text": "📲 Don't forget to forward the messages above to others and share with them!
💁 And feel free to submit your own reply if you have anything to say about this!",
          },
          "type": "template",
        },
      },
      "type": "generic",
    },
  ],
  "state": "__INIT__",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;

exports[`handles "yes" postback with other existing feedbacks 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AWDZYXxAyCdS-nWhumlz",
      "5483323992880-rumor",
      "AV-Urc0jyCdS-nWhuity",
      "AVsh8u7StKp96s659Dgq",
    ],
    "foundReplyIds": Array [
      "AWDZeeV0yCdS-nWhuml8",
    ],
    "searchedText": "貼圖",
    "selectedArticleId": "AWDZYXxAyCdS-nWhumlz",
    "selectedArticleText": "(0)(1)(/)(0)(9)(line)免費貼圖

「[全螢幕貼圖]生活市集x生活小黑熊」
https://line.me/S/sticker/10299

「松果購物 x 狗與鹿(動態貼圖) 」
https://line.me/S/sticker/10300

「LINE TV X 我的男孩限免貼圖」
https://line.me/S/sticker/10207",
    "selectedReplyId": "AWDZeeV0yCdS-nWhuml8",
  },
  "event": Object {
    "input": "y",
    "postback": Object {
      "data": "{\\"input\\":\\"y\\",\\"issuedAt\\":1519019701265}",
    },
    "timestamp": 1519019734813,
    "type": "postback",
  },
  "isSkipUser": false,
  "issuedAt": 1519019735467,
  "replies": Array [
    Object {
      "content": Object {
        "text": "We've received feedback from you and 1 other user!",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "buttons": Array [
              Object {
                "share_contents": Object {
                  "attachment": Object {
                    "payload": Object {
                      "elements": Array [
                        Object {
                          "buttons": Array [
                            Object {
                              "title": "See other replies",
                              "type": "web_url",
                              "url": "https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz",
                            },
                          ],
                          "subtitle": "Please refer to https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz for other replies to this message and references!",
                          "title": "Hey someone else says \\"(0)(1)(/)(0)(9)(line)免費貼圖

「⋯⋯ is ⚠️️ OFF TOPIC!",
                        },
                      ],
                      "template_type": "generic",
                    },
                    "type": "template",
                  },
                },
                "type": "element_share",
              },
              Object {
                "title": "Submit a new reply",
                "type": "web_url",
                "url": "https://cofacts.g0v.tw/article/AWDZYXxAyCdS-nWhumlz",
              },
            ],
            "template_type": "button",
            "text": "📲 Don't forget to forward the messages above to others and share with them!
💁 And feel free to submit your own reply if you have anything to say about this!",
          },
          "type": "template",
        },
      },
      "type": "generic",
    },
  ],
  "state": "__INIT__",
  "userId": "Uaddc74df8a3a176b901d9d648b0fc4fe",
}
`;
